_posts/2019-02-25-practice-oop-to-front-end-universal-state-module.md

```diff
@@ -0,0 +1,291 @@
+
+  @action
+  operate(item, status, state) {
+    state.goods.push(item);
+    state.status = status;
+  }
+  // call function -> this.operate({ name: 'fruits', amount: 10 }, 'open');
+}
+```
+
+### Module lifecycle
+
+`usm` provides these lifecycle events:
+
+- `moduleWillInitialize`
+- `moduleWillInitializeSuccess`
+- `moduleDidInitialize`
+- `moduleWillReset`
+- `moduleDidReset`
+
+The order in which they are run is shown in the following flow chart:
+
+![lifecycle](/integration-blog/assets/2019-02-25-practice-oop-to-front-end-universal-state-module/usm_lifecycle.png)
+
+These module lifecycles can be used to coordinate module initialization dependencies.
+
+```js
+class TodoList extends Module {  
+  async moduleWillInitialize() {
+    console.log(
+      'TodoList -> moduleWillInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+
+  async moduleWillInitializeSuccess() {
+    console.log(
+      'TodoList -> moduleWillInitializeSuccess',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+      );
+  }
+
+  async moduleDidInitialize() {
+    console.log(
+      'TodoList -> moduleDidInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+      );
+  }
+}
+
+
+class App extends Module{
+  async moduleWillInitialize() {
+    console.log(
+      'App -> moduleWillInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+
+  async moduleWillInitializeSuccess() {
+    console.log(
+      'App -> moduleWillInitializeSuccess',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+
+  async moduleDidInitialize() {
+    console.log(
+      'App -> moduleDidInitialize',
+      `pendding: ${this.pending}`, `ready: ${this.ready}`
+    );
+  }
+}
+
+const todoList = new TodoList();
+
+const app = App.create({
+  modules: {
+    todoList
+  }
+});
+```
+console.log reulst:
+
+```
+App -> moduleWillInitialize pendding: false ready: false
+TodoList -> moduleWillInitialize pendding: false ready: false
+TodoList -> moduleWillInitializeSuccess pendding: true ready: false
+App -> moduleWillInitializeSuccess pendding: true ready: false
+TodoList -> moduleDidInitialize pendding: false ready: true
+App -> moduleDidInitialize pendding: false ready: true
+```
+
+### An ideal architecture
+
+![flow chart](/integration-blog/assets/2019-02-25-practice-oop-to-front-end-universal-state-module/flow_chart.png)
+
+In a complex front-end application, this may be a more typical modular architecture design that contains the following sections:
+
+- Lifecycle
+- Store Subscriber
+- Event System
+- State
+- Dependency Modules
+- Domain Models
+
+## Conclusion
+
+When you develop using libraries or framework combinations such as React+Redux/React+MobX/Vue+Vuex and so on, it is hoped that `usm` is a nice option for modularization in your application, which may be the important modular jigsaw puzzle that you lack when building libraries such as React/Vue.
```

Comment:
Let's be concise and strong in our words:
`usm` is a module design that wants to bridge together the differences of using redux, mobx, and vuex in conjuction with different view layers such as react, vue, and angular. It is designed to help you build libraries that will work with any front-end architecture.